working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,510 @@
+  int y;
+  Point(this.x, [this.y = 0]);
+}
+
+class Point(int x, [int y = 0]);
+```
+
+Similarly for named parameters, required or not:
+
+```dart
+class Point {
+  int x;
+  int y;
+  Point(this.x, {required this.y});
+}
+
+class Point(int x, {required int y});
+```
+
+The class header can have additional elements, just like class headers
+where there is no primary constructor:
+
+```dart
+class D<TypeVariable extends Bound> extends A with M implements B, C {
+  final int x;
+  final int y;
+  const D.named(this.x, [this.y = 0]);
+}
+
+class const D.named<TypeVariable extends Bound>(int x, [int y = 0])
+    extends A with M implements B, C;
+```
+
+## Specification
+
+### Syntax
+
+The grammar is modified as follows. Note that the changes include
+support for inline classes, because they're intended to use primary
+constructors as well.
+
+```
+ <topLevelDefinition> ::=
+     <classDeclaration>
+   | <inlineClassDeclaration> // New alternative.
+   | ...;
+
+<classDeclaration> ::= // First alternative modified.
+     (<classModifiers> | <mixinClassModifiers>)
+     'class' <classNamePart> <superclass>? <interfaces>? <classBody>
+   | ...;
+
+<classNamePart> ::= // New rule.
+     'const'? <constructorName> <typeParameters>? <formalParameterList>
+   | <typeWithParameters>;
+
+<classBody> ::= // New rule.
+     '{' (<metadata> <classMemberDeclaration>)* '}'
+   | ';';
+
+<inlineClassDeclaration> ::=
+     'final'? 'inline' 'class' <classNamePart> <interfaces>? <inlineClassBody>;
+
+<inlineClassBody> ::=
+     '{' (<metadata> <inlineMemberDeclaration>)* '}'
+   | ';';
+
+<inlineMemberDeclaration> ::= <classMemberDeclaration>;
+
+<enumType> ::= // Modified rule.
+     'enum' <enumNamePart> <mixins>? <interfaces>? '{'
+        <enumEntry> (',' <enumEntry>)* (',')?
+        (';' (<metadata> <classMemberDeclaration>)*)?
+     '}';
+
+<enumNamePart> ::= // New rule.
+     <constructorName> <typeParameters>? <formalParameterList>
+   | <typeWithParameters>;
+```
+
+The word `inline` is now used in the grammar, but it is not a reserved word
+or a built-in identifier.
+
+### Static processing
+
+A class declaration with a primary constructor *(including `inline`, but
+not `<mixinApplicationClass>`, because the grammar rules do not include
+those)* is desugared to a class declaration without a primary
+constructor. An enum declaration with a primary constructor is desugared
+using the same steps. This determines the static analysis and dynamic
+semantics of a primary constructor.
+
+*In other words, there is no other semantics than the desugaring. When
+desugared, the resulting class declaration is treated exactly the same as
+it would have been if the developer had written the result of the
+desugaring step in the first place.*
+
+A compile-time error occurs if a formal parameter that contains a
+term of the form `this.v`, or `super.v` where `v` is an identifier has the
+modifier `covariant`. *For example, `required covariant int this.v` is an
+error.*
```

Comment:
Og destinationer only determines the dynamic semantics, then we need to specify the complete static semantics of the primary constructor syntax. That's still tricky, because it includes complete type inference, but at least for a fairly limited syntax. We're probably able to reuse the type inference of the parameter list from the analysis of other constructors, but we still have to check that any place where we say that "class declares a constructor" or "class declares an instance variable", it also includes a class with a primary constructor. We cannot assume that a declared variable has a particular declaring syntax any more.

So, basically, we have to say, during static analysis, that a class using the primary constructor syntax _declares_ instance variables and a constructor, and specify their declared types, finality, covariance, and the constructor's function type, which is inferred by type inference like other constructors.

Doable. But gotta keep the i's dotted and t's crossed all the way through.

The sooner we move away from the physical syntax, the sooner we can unify different ways to specify the same thing.